src/data/equiv/basic.lean

```diff
@@ -563,26 +563,68 @@ subtype_congr (equiv.refl α) (assume a, h ▸ iff.refl _)
 def set_congr {α : Type*} {s t : set α} (h : s = t) : s ≃ t :=
 subtype_congr_prop h
 
-def subtype_subtype_equiv_subtype {α : Type u} (p : α → Prop) (q : subtype p → Prop) :
+def subtype_subtype_equiv_subtype_ex {α : Type u} (p : α → Prop) (q : subtype p → Prop) :
   subtype q ≃ {a : α // ∃h:p a, q ⟨a, h⟩ } :=
 ⟨λ⟨⟨a, ha⟩, ha'⟩, ⟨a, ha, ha'⟩,
   λ⟨a, ha⟩, ⟨⟨a, ha.cases_on $ assume h _, h⟩, by { cases ha, exact ha_h }⟩,
   assume ⟨⟨a, ha⟩, h⟩, rfl, assume ⟨a, h₁, h₂⟩, rfl⟩
 
-/- A subtype of a sigma-type is a sigma-type over a subtype. -/
+def subtype_subtype_equiv_subtype_inter {α : Type u} (p q : α → Prop) :
+  {x : subtype p // q x.1} ≃ subtype (λ x, p x ∧ q x) :=
+(subtype_subtype_equiv_subtype_ex p _).trans $
+subtype_congr_right $ λ x, exists_prop
+
+/-- If the outer subtype has more restrictive predicate than the inner one,
+then we can drop the latter. -/
+def subtype_subtype_equiv_subtype {α : Type u} {p q : α → Prop} (h : ∀ {x}, q x → p x) :
+  {x : subtype p // q x.1} ≃ subtype q :=
+(subtype_subtype_equiv_subtype_inter p _).trans $
+subtype_congr_right $
+assume x,
+⟨and.right, λ h₁, ⟨h h₁, h₁⟩⟩
+
+/-- If a proposition holds for all elements, then the subtype is
+equivalent to the original type. -/
+def subtype_univ_equiv {α : Type u} {p : α → Prop} (h : ∀ x, p x) :
+  subtype p ≃ α :=
+⟨λ x, x, λ x, ⟨x, h x⟩, λ x, subtype.eq rfl, λ x, rfl⟩
+
+/-- A subtype of a sigma-type is a sigma-type over a subtype. -/
 def subtype_sigma_equiv {α : Type u} (p : α → Type v) (q : α → Prop) :
   { y : sigma p // q y.1 } ≃ Σ(x : subtype q), p x.1 :=
-begin
-  fsplit,
-  rintro ⟨⟨x, y⟩, z⟩, exact ⟨⟨x, z⟩, y⟩,
-  rintro ⟨⟨x, y⟩, z⟩, exact ⟨⟨x, z⟩, y⟩,
-  rintro ⟨⟨x, y⟩, z⟩, refl,
-  rintro ⟨⟨x, y⟩, z⟩, refl,
-end
+⟨λ x, ⟨⟨x.1.1, x.2⟩, x.1.2⟩,
+ λ x, ⟨⟨x.1.1, x.2⟩, x.1.2⟩,
+ λ ⟨⟨x, h⟩, y⟩, rfl,
+ λ ⟨⟨x, y⟩, h⟩, rfl⟩
+
+/-- A sigma type over a subtype is equivalent to the sigma set over the original type,
+if the fiber if empty outside of the subset -/
+def sigma_subtype_equiv_of_subset {α : Type u} (p : α → Type v) (q : α → Prop)
+  (h : ∀ x (y : p x), q x) :
+  (Σ x : subtype q, p x) ≃ Σ x : α, p x :=
+(subtype_sigma_equiv p q).symm.trans $ subtype_univ_equiv $ λ x, h x.1 x.2
+
+def sigma_subtype_preimage_equiv {α : Type u} {β : Type v} (f : α → β) (p : β → Prop)
+  (h : ∀ x, p (f x)) :
+  (Σ y : subtype p, {x : α // f x = y}) ≃ α :=
+calc _ ≃ Σ y : β, {x : α // f x = y} : sigma_subtype_equiv_of_subset _ p (λ y ⟨x, h'⟩, h' ▸ h x)
+   ... ≃ α                           : sigma_preimage_equiv f
+
+def sigma_subtype_preimage_equiv_subtype {α : Type u} {β : Type v} (f : α → β)
+  {p : α → Prop} {q : β → Prop} (h : ∀ x, p x ↔ q (f x)) :
+  (Σ y : subtype q, {x : α // f x = y}) ≃ subtype p :=
+calc (Σ y : subtype q, {x : α // f x = y}) ≃
+  Σ y : subtype q, {x : subtype p // subtype.mk (f x) ((h x).1 x.2) = y} :
+  begin
+  apply sigma_congr_right,
```

Comment:
Please indent the contents of the `begin`-`end`-block

Reply:
Done.